### DISCUSSION TITLE
[Feature] Duplicate Detector - Synchronise metadata and album membership between duplicate images

### DISCUSSION THREAD

**Discussion by chrisscotland** [immich-app/immich#10262] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

The new image (de)duplication feature is great but sorting out the duplicates it identifies can be hard.

I have many duplicate images where one version is added to an album and the other is not. Sometimes the copy I wish to keep is not added to an album and by trashing the other copy the album is then missing an image. Adding the replacement image back in to the album(s) manually is labour intensive - also see suggestion #10206.

More generally it would be nice to have the ability to synchronise the metadata between the two images before selecting one to delete. One image may have manually updated geolocation, date/time, description and album membership which I would like to copy to the version being kept so this information is retained.

![image](https://github.com/immich-app/immich/assets/142149847/575e07fb-ed97-4cfa-9446-99e40c9be1b5)

### Platform

- [ ] Server
- [x] Web
- [ ] Mobile

**Comment by antonio-ivanovski** [comment]
I am facing similar issue. I have bunch of videos that were imported from Google Photos but also have had those videos in their original state. I would like to keep the original ones but while doing so not lose the album state of those videos.

The immich-go duplicate command did nice job for all the images, but for the videos it didn't work that well, maybe give a try if only issue you have is images: https://github.com/simulot/immich-go?tab=readme-ov-file#command-duplicate

<img width="535" alt="image" src="https://github.com/immich-app/immich/assets/1628876/b72966e2-4bad-485f-8706-af7fe8d8079c">

**Comment by di5cord20** [comment]
I just ran into the same scenario.  Would very much welcome the ability to keep/copy metadata from the trashed version and store in the kept version.

**Comment by That-Guy-Jack** [comment]
Also want to add to this, i have a bunch of photos from google that i originally imported before i included albums, images are basically the same but some are in albums and some aren't

**Comment by hstock** [comment]
It would also be nice to be able to delete some and create a stash for others. I might have a degraded copy from am messengin app, an original JPG and an original RAW. Or it might detect some similar but not equal images and some lower quality copies as the same image.

**Comment by R-Rudolf** [comment]
I did a small search on the topic, regarding alternatives:

**Immich-go** duplicate finder is based on filenames and dates, in contrast to the new immich native duplicate finder based on content similarity, which is more robust. Next to it, does not handle metadata, only harmonizes album membership after deduplication, see [code](https://github.com/simulot/immich-go/blob/7ee54ba2b7b7f84b2e2e2ebc5c15de8bb5383bac/cmd/duplicate/duplicate.go#L162C8-L169C9).

The other alternative: [**immich_duplicate_finder**](https://github.com/vale46n1/immich_duplicate_finder/) is doing content-based duplicate finding, but that has simple delete logic ([code link](https://github.com/vale46n1/immich_duplicate_finder/blob/6d985ae8b995c3d64e4befe2fcc19f7ebd01ccae/utility.py#L44-L46)), nothing else.

**Comment by sam-dumont** [comment]
This feature would be very handy on iOS : if you choose to synchronise shared albums, you end up with a duplicated degraded version in the Immich album (unfortunately, iOS photos stores the shared album version in a lower resolution). So I currently I need to do a lot of manual step to keep the higher res in an Immich album. 

Being able to 
1) mark as duplicate
2) keep the high resolution asset in both albums
3) exclude the lower resolution in future album syncs so it does not upload again

would be amazing

**Comment by EinToni** [comment]
Just as a info, I started on the implementation on this: https://github.com/immich-app/immich/pull/13851
If you have improvement ideas let me know here or in the PR.

**Reply by chrisscotland** [reply]
Amazing, this looks really promising!

**Reply by lory9995** [reply]
Thanks you! i was migrating and then i found the non-sharing exif data in my duplicates and stopped right away. Absolutely need this function, it's practically a must-have-basics kind. 
Glad someone is doing it, thanks you so much!

**Comment by lory9995** [comment]
If i can add something, i would love if it would be possible to choose, maybe in settings, to choose a preferred format ( like between .heic and jpg )

**Comment by ZuluForce** [comment]
A quick and dirty script to synchronize albums but nothing else. #13851 looks very promising but I knew for my situation I just needed albums to sync to get on with the rest of my cleanup.

```
import requests
import json

immich_api_base = "<immich_host_port>/api"
api_key = '<your_api_key>'

headers = {
  'Content-Type': 'application/json',
  'Accept': 'application/json',
  'x-api-key': api_key
}

duplicates = requests.get(f"{immich_api_base}/duplicates", headers=headers, data={}).json()

for dup in duplicates:
	assets = dup['assets']

	# Union of album ids and names for all assets in this duplicate
	all_albums = set()
	asset_ids = [a['id'] for a in assets]
	for asset_id in asset_ids:
		params = {'assetId': asset_id}
		albums_resp = requests.get(f"{immich_api_base}/albums", headers=headers, params=params)
		albums_resp.raise_for_status()

		resp_json = albums_resp.json()

		all_albums.update([(a['id'], a['albumName']) for a in resp_json])

	for album_id, album_name in all_albums:
		print(f"Adding assets {asset_ids} to album '{album_name}' (id: {album_id})")
		payload = {'ids': asset_ids}
		add_to_album_resp = requests.put(f"{immich_api_base}/albums/{album_id}/assets", headers=headers, json=payload)
		add_to_album_resp.raise_for_status()
```

Take it for what it is. There's basically no error handling except to give up right away. In my several thousand dupes that needed syncing it had no problems. YMMV

**Reply by timonrieger** [reply]
dropping in [immich-py](https://github.com/timonrieger/immich-py) here as an alternative way to implement the script 😄

**Reply by Tr4ansp0rter** [reply]
exactly what I have been looking for. Thanks a lot!

**Comment by hoaluvn** [comment]
Yes, it's great to have capacity to set rule for delete or keep

**Comment by kacboy** [comment]
https://github.com/richardjlyon/immich-lib

Someone made a standalone tool to dedup and sync metadata (but not albums) if anyone is interested. Maybe with that album script above, this will get some folks all the way there :)

**Comment by chrisscotland** [comment]
This feature is now available as of v2.7.0 

https://github.com/immich-app/immich/pull/25316
